core/src/main/scala/scalan/primitives/Structs.scala

```diff
@@ -202,33 +222,35 @@ trait Structs extends StructTags { self: Scalan =>
 
   val flatteningBuilder = new IsoBuilder { def apply[S](e: Elem[S]) = getFlatteningIso(e) }
 
-  def flatteningIso[T](eTo: StructElem[T]): Iso[_,T] = {
-    val flatIsos = eTo.fields.collect {
-      case (fn, fe: StructElem[_]) => (fn, flatteningIso(fe).asInstanceOf[Iso[Any,Any]])
-      }.toMap
-
-    if (flatIsos.isEmpty) return identityIso(eTo)
+  def flatteningIso[T <: Struct](eTo: StructElem[T]): Iso[_, T] = {
+    val flatIsos: Map[String, Iso[_, _]] = eTo.fields.collect {
+      case (fn, fe: StructElem[_]) => (fn, flatteningIso(fe))
+    }.toMap
+
+    if (flatIsos.isEmpty) {
+      identityIso(eTo)
+    } else {
+
+      // relate resulting field types by original field name
+      val fromFields = eTo.fields.flatMap {
+        case (fn, fe) =>
+          flatIsos.get(fn) match {
+            case Some(iso) =>
+              iso.eFrom match {
+                case flatElem: StructElem[_] =>
+                  flatElem.fields.map { case (nestedName, nestedE) => (fn, nestedName -> nestedE) }
+                case _ => !!!(s"StructElem is expected as eFrom of flattened Iso $iso")
+              }
+            case None => List((fn, "" -> fe))
+          }
+      }
 
```

Comment:
I don't think it's more readable and we don't use the idiom elsewhere (looking through codebase). I can restore if you prefer it.